src/asymmetric_crypto/ecies.rs

```diff
@@ -0,0 +1,313 @@
+        + Sub<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Mul<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Div<&'b DH::PrivateKey, Output = DH::PrivateKey>,
+{
+    // Extract the sender's ephemeral public key R from the ciphertext
+    let ephemeral_public_key = &ciphertext[..PUBLIC_KEY_LENGTH];
+    let sender_public_key = DH::PublicKey::try_from_bytes(ephemeral_public_key)?;
+
+    // Calculate the shared secret point (Px, Py) = P = R.y = r.G.y = r.Y
+    let shared: DH::PublicKey = &sender_public_key * receiver_private_key;
+
+    // S. Note: ECIES formally uses S = Px rather than the serialization of P
+    let mut shared_bytes = shared.to_bytes().to_vec();
+    // S||S1: if the user provided shared_encapsulation_data S1, then we append it
+    // to the shared_bytes S
+    if let Some(s1) = shared_encapsulation_data {
+        shared_bytes.extend(s1);
+    }
+
+    // Generate the 256-bit symmetric decryption key k, derived using Shake128 XOF
+    const SYMMETRIC_KEY_LENGTH: usize = Aes256GcmCrypto::KEY_LENGTH;
+    let key = kdf!(SYMMETRIC_KEY_LENGTH, &shared_bytes);
+
+    // Extract the nonce from the ciphertext
+    let nonce_start = PUBLIC_KEY_LENGTH;
+    let nonce_end = nonce_start + <Aes256GcmCrypto as Dem<SYMMETRIC_KEY_LENGTH>>::Nonce::LENGTH;
+    let nonce = &ciphertext[nonce_start..nonce_end];
+
+    // Separate the encrypted message and MAC from the ciphertext
+    let c_d = &ciphertext[nonce_end..];
+
+    // Decrypt and verify the message using AES-256-GCM
+    let decrypted_message = decrypt_combined(&key, c_d, nonce, shared_authentication_data)?;
+
+    Ok(decrypted_message)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{ecies_decrypt, ecies_encrypt, CryptoCoreError};
+    use crate::{
+        asymmetric_crypto::{curve25519::X25519KeyPair, DhKeyPair},
+        reexport::rand_core::SeedableRng,
+        CsRng,
+    };
+
+    #[test]
+    fn test_encrypt_decrypt() -> Result<(), CryptoCoreError> {
+        let mut rng = CsRng::from_entropy();
+        let key_pair: X25519KeyPair = X25519KeyPair::new(&mut rng);
+        let msg = b"Hello, World!";
+
+        // Encrypt the message
+        let encrypted_message = ecies_encrypt::<
+            CsRng,
+            X25519KeyPair,
+            { X25519KeyPair::PUBLIC_KEY_LENGTH },
+            { X25519KeyPair::PRIVATE_KEY_LENGTH },
+        >(&mut rng, key_pair.public_key(), msg, None, None)?;
+
+        // Decrypt the message
+        let decrypted_message = ecies_decrypt::<
+            X25519KeyPair,
+            { X25519KeyPair::PUBLIC_KEY_LENGTH },
+            { X25519KeyPair::PRIVATE_KEY_LENGTH },
+        >(key_pair.private_key(), &encrypted_message, None, None)?;
+
+        // Check if the decrypted message is the same as the original message
+        assert_eq!(msg, &decrypted_message[..]);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_encrypt_decrypt_with_optional_data() -> Result<(), CryptoCoreError> {
```

Comment:
Maybe you can check you cannot decrypt if both parties are not using the same `authentication_data` or `encapsulation_data`.